src/util-lua-tls.c

```diff
@@ -316,48 +264,111 @@ static int GetCertChain(lua_State *luastate, const Flow *f, int direction)
+static int LuaTlsGetClientCertChain(lua_State *luastate)
+{
+    return GetCertChain(luastate, true);
+}
 
-    int direction = LuaStateGetDirection(luastate);
+static int GetCertSerial(lua_State *luastate, bool client)
+{
+    struct LuaTls *s = (struct LuaTls *)luaL_checkudata(luastate, 1, tls_state_mt);
+    if (s->state == NULL) {
+        LUA_ERROR("failed to get flow");
+    }
 
-    Flow *f = LuaStateGetFlow(luastate);
-    if (f == NULL)
-        return LuaCallbackError(luastate, "internal error: no flow");
+    const SSLStateConnp *connp;
 
-    r = GetCertChain(luastate, f, direction);
+    if (client) {
+        connp = &s->state->client_connp;
+    } else {
+        connp = &s->state->server_connp;
+    }
+    if (connp->cert0_serial == NULL)
+        return LuaCallbackError(luastate, "error: no certificate serial");
 
-    return r;
+    return LuaPushStringBuffer(
+            luastate, (uint8_t *)connp->cert0_serial, strlen(connp->cert0_serial));
 }
 
-/** \brief register tls lua extensions in a luastate */
-int LuaRegisterTlsFunctions(lua_State *luastate)
+static int LuaTlsGetServerCertSerial(lua_State *luastate)
 {
-    /* registration of the callbacks */
-    lua_pushcfunction(luastate, TlsGetCertNotBefore);
-    lua_setglobal(luastate, "TlsGetCertNotBefore");
+    return GetCertSerial(luastate, false);
+}
 
-    lua_pushcfunction(luastate, TlsGetCertNotAfter);
-    lua_setglobal(luastate, "TlsGetCertNotAfter");
+static int LuaTlsGetClientCertSerial(lua_State *luastate)
+{
+    return GetCertSerial(luastate, true);
+}
 
-    lua_pushcfunction(luastate, TlsGetVersion);
-    lua_setglobal(luastate, "TlsGetVersion");
+static int GetAgreedVersion(lua_State *luastate, bool client)
+{
+    struct LuaTls *s = (struct LuaTls *)luaL_checkudata(luastate, 1, tls_state_mt);
+    if (s->state == NULL) {
+        LUA_ERROR("failed to get state");
+    }
 
-    lua_pushcfunction(luastate, TlsGetCertInfo);
-    lua_setglobal(luastate, "TlsGetCertInfo");
+    uint16_t version;
+    if (client) {
+        version = s->state->client_connp.version;
+    } else {
+        version = s->state->server_connp.version;
+    }
 
-    lua_pushcfunction(luastate, TlsGetSNI);
-    lua_setglobal(luastate, "TlsGetSNI");
+    char ssl_version[SSL_VERSION_MAX_STRLEN];
+    SSLVersionToString(version, ssl_version);
 
-    lua_pushcfunction(luastate, TlsGetCertSerial);
-    lua_setglobal(luastate, "TlsGetCertSerial");
+    lua_pushstring(luastate, (const char *)&ssl_version);
+    return 1;
+}
 
-    lua_pushcfunction(luastate, TlsGetCertChain);
-    lua_setglobal(luastate, "TlsGetCertChain");
+static int LuaTlsGetServerVersion(lua_State *luastate)
+{
+    return GetAgreedVersion(luastate, false);
+}
 
-    return 0;
+static int LuaTlsGetClientVersion(lua_State *luastate)
+{
+    return GetAgreedVersion(luastate, true);
+}
+
+static const struct luaL_Reg tlslib_meta[] = { { "get_server_cert_not_before",
```

Comment:
formatting looks a bit wonky here. I think other libs have a clang-format exception and use manual formatting

Reply:
Will adjust